### DISCUSSION TITLE
Controlling Floppy Drives Directly From MIDI Controller (Keyboard etc.)

### DISCUSSION THREAD

**Discussion by acarreisli** [SammyIAm/Moppy2#186] [Q&A]
Hi,
How can I control floppy drives from a MIDI controller (MIDI keyboard) without a computer? Is it possible with Moppy2?

**Comment by DJthefirst** [comment]
So Moppy uses a proprietary protocol when sending messages between the Microcontroller and Arduino. Which makes it hard to control Moppy without a computer. The second problem is all the mappers used to control the logic of what note is playing is server side. Without fundamentally changing moppy you can't use it without the server. 

The best way is probably to uses Moppys Midi in functionality but that requires a computer. Are you using midi Din(OG Midi) or Midi Usb? You could use another microcontroller as an interpreter and write a script to convert the data. You could also get something like a LatePanda to run Moppy for you if you have the money. 

I've written my own program similar to Moppy but adhering to the MIDI standard and with distributors the counter part to mappers that run directly on the microcontroller. I haven't tested it without a computer yet but I designed it to work without a server. Only Problem is I only support ESP32 Atm 😕.. I can let you know when I add other platforms if your interested but I probably won't get around to it until this summer.

**Reply by acarreisli** [reply]
I'm new on floppy music world :) I've set up and tried Moppy recently and it have more functions than I think. I've tried MIDI connection directly from a DAW (via loopMIDI software). Currently I don't have a MIDI keyboard (with a 5-pin DIN MIDI output) but I want to get one. I have a NodeMCU ESP-32S. If your software runs on it, I want to try it.

**Reply by DJthefirst** [reply]
That Esp32 should work just fine with the software. It's still a work in progress so ill see if I can get things together this weekend. What type of setup are you doing? The esp32 can only support about 7-8 Floppy drives as it has 15 DIO pins. It has a few others but they are multipurpose so you might be able to get 9 drives but I haven't tested it. I've mostly used other instruments so ill have to test the code again with floppy drives. The instruments I support are Drumkits, Stepper Motor Drivers, and PWM Outputs for other devices. Some examples can be seen on my [youtube channel](https://www.youtube.com/@DJthefirst/videos).  If you want the Midi Din connector input I can work on the code for it but for now, it is just USB Serial. 

 I will also try to see if I can figure out a direct USB Midi Keyboard to Esp Serial as both are Client Devices and Depend on the Computer to act as a USB host. As the program takes in direct midi you can just use loop midi without the need for a GUI. I have two AKAI keyboards I use with my instruments through loop Midi. I also have a USB to Midi Din connector for my tesla coils but I find it tedious to wire the esp for it. 

Is there a reason you want to go computerless? I also just thought you might be able to use a Raspberry Pi as the server. Sammy would know better if others have used it with moppy. My software should also work on Rpi and I think loop midi has Linux support. 

My software is stable but still a work in progress. I'll see what I can do this weekend and I also have to move floppy drives out of Dev. here is a link to the main [GitHub](https://github.com/DJthefirst/Mechanical-Midi-Music). It is just a PlatformIO project you can load into VSCode. Much better than the Arduino IDE. The configurations are in the main.cpp where you can create Distributors like Midi mappers and Device.h where you can set the config. Pinout is listed in Platform.h where pairs of pins in the array would represent the Direction, Step connections. I'll let you know when I make the changes if you are still interested. If not you should check back later this summer when I can get allot more work done. I'm in college atm so classes take up most of my time.

**Reply by samphonic** [reply]
> Is there a reason you want to go computerless? I also just thought you might be able to use a Raspberry Pi as the server. Sammy would know better if others have used it with moppy.

Raspberry Pi is definitely a potential option, though I haven't seen a complete command-line control option for Moppy anywhere yet, so you'd still have to connect the RPi (at least to get things setup) to a screen of some kind.  @DJthefirst 's loop midi option with his software on a Raspberry Pi might be a better headless option since it could just run on startup.

Not a lot to add to what DJ has already said: The MIDI messages are converted to MoppyMessages in the Java program, so that Java program needs to be run somewhere for Moppy to work.  This isn't a *technical* limitation so much as Moppy was my first foray into C code so I didn't know how to efficiently get all of that working on the same microcontroller.  DJ's software is probably closer to being able to do what you're looking for, so that's likely the best avenue if you need something without a PC.

**Reply by DJthefirst** [reply]
Just a quick note here. I still have 4 more weeks of my internship before I can start grinding on the code but the esp32 S2,S3 have USB OTG support. This means once I clean up the core I can make a branch off main where no midi usb to serial converter is needed and the ESP will act like a normal midi usb interface. 100% headless.
